### PR TITLE
Categories hierarchy : Fix #411

### LINKED ISSUE #411: [Categorization] page - display category hierachies
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am running the latest version of the COMET-WEB
- [x] I have searched [open](https://github.com/RHEAGROUP/COMET-WEB-Community-Edition/issues) and [closed](https://github.com/RHEAGROUP/COMET-WEB-Community-Edition/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description

1. Add a new page in the RDL part in which categories can be visualized and created/edited/et
2. The page needs to show a list of categories in the left (might need a datagrid to show shortname, name, and container RDL)
3. Show a z-diagram based panel that diplays the generalization hierachy of a selected Category (both super and sub categories taking multiple inheritance into account). The selection should be more than a click on the category, see note below

 - For the future, i foresee that users need to be able to drag-n-drop categories onto the canvas and create a link between them that states the generalization as per 10-25 (preferably using UML semantics when drawing the relationship and arrows)
 - For the future, i foresee that from the list a user will need to be able to edit categories (either using a master/detail structure or popu dialog)

### Steps to Reproduce
NA

### Logs
NA  
### System Configuration

- COMET version:
  - [ ] COMET-WEB: 1.1.0.0
  - [ ] Other: 
- Web Services:
  - [ ] COMET Web Services version:

### PR DESCRIPTION
**Pull request by nabilatrhea**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/COMET-WEB-Community-Edition/pulls) open
- [x] I have verified that I am following the COMET-WEB [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/COMET-WEB-Community-Edition/master/.github/CONTRIBUTING.md)
- [x] I have provided test coverage for my change (where applicable)

### Description

Fix #411 
Display categorties hierarchy using z-diagram (needs some adjustments to display the diagram on the right of the datagrid)
 
image

### CODE REVIEW

**Comment by samatrhea** [comment]
please use semantics from UML (should have been more explicit in the original feature request)

the direction of the arrow is from subclass to superclass and the arrow is an open arrow, see example image below (the "extends"word does not need to be in the arrow)

image

Some more requests: in case the Category IsAbstarct, set the caption to Italics

**Comment by nabilatrhea** [comment]
@samatrhea Yes, i have added the Open arrow. however it is not displayed (COMETwebapp/wwwroot/images/UML_Open_arrow.svg) i will try to fix that and add the Italic caption for the abstract categories

**Review by samatrhea — CHANGES_REQUESTED** [review]

File: COMETwebapp/ViewModels/Components/ReferenceData/CategoriesTableViewModel.cs [hunk 1]
```diff
@@ -408,5 +415,40 @@ private void UpdateRows(IEnumerable<CategoryRowViewModel> rowsToDisplay)
...
+
+            this.CategoryHierarchyDiagramViewModel.SetupDiagram();
+        }
+
+        /// <summary>
+		/// Get the subcategories of a category
+		/// </summary>
+		/// <param name="category">The category
+        /// <returns>A <see cref="List{Category}" /></returns>
+        public List<Category> GetSubCategories(Category category)
```
[the remark this answers is not in the document]
Comment (nabilatrhea) [reply COMETwebapp/ViewModels/Components/ReferenceData/CategoriesTableViewModel.cs:439]: Thanks !

**Review by samatrhea — CHANGES_REQUESTED** [review]
I have the impression that in the current solution the arrows are pointing in the wrong direction. The direction should be from subclass to superclass: subclass ---> superclass; please have a look at the diagram i provided as an example

**Comment by nabilatrhea** [comment]
@samatrhea 
image
image
image

File: COMETwebapp/wwwroot/css/app.css
Comment (jaimeatrhea) [inline COMETwebapp/wwwroot/css/app.css:1]: Why does this file contain the css of the CategoryNodeComponent? I think that should be in the component

**Review by jaimeatrhea — CHANGES_REQUESTED** [review]

File: COMETwebapp/wwwroot/css/app.css
Comment (nabilatrhea) [reply COMETwebapp/wwwroot/css/app.css:1]: I tough the css isolation isn't working when deploying the app

Comment (jaimeatrhea) [reply COMETwebapp/wwwroot/css/app.css:1]: Hmmm never heard of that, but if the rest of the components have css isolated it seems like it's working

**Comment by samatrhea** [comment]
> @samatrhea image image image

ok,. so the direction is then indeed correct, its is typical to have the most abstract class on top and the subclasses below. The diagram reads a bit contra intuitiv -> it needs to be flipped

File: COMETwebapp/Components/ReferenceData/CategoryNodeComponent.razor.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+//    The COMET WEB Community Edition is distributed in the hope that it will be useful,
+//    but WITHOUT ANY WARRANTY; without even the implied warranty of
+//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+//    Affero General Public License for more details.
+//
+//    You should have received a copy of the GNU Affero General Public License
+//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+using Microsoft.AspNetCore.Components;
```
Comment (samatrhea) [inline COMETwebapp/Components/ReferenceData/CategoryNodeComponent.razor.cs:24]: using statments go inside namespace

**Review by samatrhea — CHANGES_REQUESTED** [review]
minor editorial change of formatting

**Review by samatrhea — APPROVED** [review]

File: COMETwebapp/ViewModels/Components/ReferenceData/CategoriesTableViewModel.cs [hunk 3]
```diff
@@ -110,6 +112,11 @@ public CategoriesTableViewModel(ISessionService sessionService, IShowHideDepreca
         /// </summary>
         public IShowHideDeprecatedThingsService ShowHideDeprecatedThingsService { get; }
 
+        /// <summary>
```
Comment (antoineatrhea) [inline COMETwebapp/ViewModels/Components/ReferenceData/CategoriesTableViewModel.cs:115]: Indentation problem inside the file

File: COMETwebapp/Components/ReferenceData/CategoryHierarchyDiagram.razor.cs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    /// </summary>
+    public partial class CategoryHierarchyDiagram
+    {
+        /// <summary>
+        ///     The <see cref="ICategoryHierarchyDiagramViewModel" /> for the component
+        /// </summary>
+        [Parameter]
+        public ICategoryHierarchyDiagramViewModel ViewModel { get; set; }
+
+        protected override void OnInitialized()
```
Comment (antoineatrhea) [inline COMETwebapp/Components/ReferenceData/CategoryHierarchyDiagram.razor.cs:45]: Missing documentation

File: COMETwebapp/ViewModels/Components/ReferenceData/CategoryHierarchyDiagramViewModel.cs [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+        /// <summary>
+        /// The categories hierarchy <see cref="Diagram" /> to display
+        /// </summary>
+        public Diagram Diagram { get; set; }
+
+        /// <summary>
+        /// The svg path of arrow
+        /// </summary>
+        const string svgArrowPath = "M -0.093 17.86 V -18.5 L 29.233 -0.32 L -0.093 17.86 z M 1.407 -15.806 v 30.9715 L 26.3865 -0.32 L 1.407 -15.806 z";
```
Comment (antoineatrhea) [inline COMETwebapp/ViewModels/Components/ReferenceData/CategoryHierarchyDiagramViewModel.cs:74]: You can maybe create a SVG file with that contains and use it as a css class?

File: COMETwebapp/ViewModels/Components/ReferenceData/ICategoriesTableViewModel.cs [hunk 6]
```diff
@@ -95,6 +95,11 @@ public interface ICategoriesTableViewModel : ISingleIterationApplicationBaseView
         /// </summary>
         ReferenceDataLibrary SelectedReferenceDataLibrary { get; set; }
 
+        /// <summary>
```
Comment (antoineatrhea) [inline COMETwebapp/ViewModels/Components/ReferenceData/ICategoriesTableViewModel.cs:98]: Indentation problem inside the file

**Review by antoineatrhea — APPROVED** [review]
Some small editorial problem and a possible improvement.
Global question: What happen on the diagram if one of the category in the category hierarchy is deprecated?

File: COMETwebapp/ViewModels/Components/ReferenceData/CategoryHierarchyDiagramViewModel.cs (hunk 5 above)
Comment (nabilatrhea) [reply COMETwebapp/ViewModels/Components/ReferenceData/CategoryHierarchyDiagramViewModel.cs:74]: We can't do that, the library uses svg path to create a LinkMarker

**Comment by nabilatrhea** [comment]
> Some small editorial problem and a possible improvement. Global question: What happen on the diagram if one of the category in the category hierarchy is deprecated?

Now if a category is deprecated, the diagram is updated also on session refresh

**Pull request merged**